Implement traceable torch.tensor when you have SymInt/SymFloat inputs

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109515
* #109513

I just ported the C++ torch.tensor implementation to Python, swapping out the inner bits to successively stack tensors together, so that we can trace through `scalar_tensor`.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang committed:
Implement traceable torch.tensor when you have SymInt/SymFloat inputs

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Implement traceable torch.tensor when you have SymInt/SymFloat inputs"


I just ported the C++ torch.tensor implementation to Python, swapping out the inner bits to successively stack tensors together, so that we can trace through `scalar_tensor`.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
 
+
+def _isStorage(obj):
+    # TODO: this is slightly inaccurate; we don't allow subclasses on
+    # UntypedStorage
```
why not type check directly?

ezyang replied:
Hmm, so, originally I was Dynamo'ing through this code and Dynamo didn't support the type() is ... test. But then I realized I didn't need to Dynamo through it I could make_fx trace it instead, probably can fix it now

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+        )
+    type_inference = dtype is None
+    new_tensor = _internal_new_from_data(
+        {"device": "cpu"},  # TODO: use torch.get_default_tensor_type
```
why device cpu??

ezyang replied:
Because that's the default device, when you don't specify device

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
 zero_ = _make_inplace(zero)
 
+
+def _isStorage(obj):
```
hypernit camelcase

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+        )
+        var = data
+        if copy_variables:
+            var = var.detach()
```
sorry, why detach and not clone?

ezyang replied:
you know, I have no idea why the og code does that.  The logic was added in https://github.com/pytorch/pytorch/pull/14097 but we didn't discuss it in review. If I had to guess, it is to deal with the situation if var requires_grad=True, on the inside we might do a .to() call and we want to avoid generating a backwards graph when this happens.

voznesenskym reviewed (approved):
stamp because port but questions

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+            if cur_item is obj:
+                raise TypeError("new(): self-referential lists are incompatible")
+            """
+            item_scalarType = _infer_scalar_type(cur_item)  # recurse!
```
I know its ported but camelcase

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+            """
+            item_scalarType = _infer_scalar_type(cur_item)  # recurse!
+            if scalarType is not None:
+                scalarType = torch.promote_types(scalarType, item_scalarType)
```
more camels

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+
+
+# Analogous to recursive_store
+def _recursive_build(sizes, dim, scalarType, obj):
```
assert dim < sizes?

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+    type_inference = dtype is None
+    new_tensor = _internal_new_from_data(
+        {"device": "cpu"},  # TODO: use torch.get_default_tensor_type
+        dtype if dtype is not None else torch.get_default_dtype(),
```
nit: type_inference bool instead of `dtype is not None`

voznesenskym commented on torch/_refs/__init__.py:
```diff
@@ -6128,6 +6128,184 @@ def vdot(self, other):
+        var = data
+        if copy_variables:
+            var = var.detach()
+        inferred_scalar_type = var.dtype if type_inference else scalar_type
```
this is the same line as `inferred_scalar_type = _infer_scalar_type(data) if type_inference else scalar_type` below because that line routes to `.dtype` anyway - combine and move up?

ezyang commented:
I resolved all the PR comments that would not have caused divergence from the original code

ezyang committed:
Update on "Implement traceable torch.tensor when you have SymInt/SymFloat inputs"


I just ported the C++ torch.tensor implementation to Python, swapping out the inner bits to successively stack tensors together, so that we can trace through `scalar_tensor`.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "known master breakage only"
